mesonbuild/modules/simd.py

```diff
@@ -13,76 +13,110 @@
 # limitations under the License.
 from __future__ import annotations
 
-from .. import mesonlib, compilers, mlog
+import typing as T
+
+from .. import mesonlib, mlog
 from .. import build
+from ..compilers import Compiler
+from ..interpreter.type_checking import SOURCES_KW
+from ..interpreterbase.decorators import KwargInfo, typed_pos_args, typed_kwargs
 
 from . import ExtensionModule, ModuleInfo
 
+if T.TYPE_CHECKING:
+    from typing_extensions import TypedDict
+
+    from . import ModuleState
+    from ..interpreter import Interpreter
+
+    class CheckKw(TypedDict):
+
+        compiler: Compiler
+        mmx: T.List[mesonlib.FileOrString]
+        sse: T.List[mesonlib.FileOrString]
+        sse2: T.List[mesonlib.FileOrString]
+        sse3: T.List[mesonlib.FileOrString]
+        ssse3: T.List[mesonlib.FileOrString]
+        sse41: T.List[mesonlib.FileOrString]
+        sse42: T.List[mesonlib.FileOrString]
+        avx: T.List[mesonlib.FileOrString]
+        avx2: T.List[mesonlib.FileOrString]
+        neon: T.List[mesonlib.FileOrString]
+
+
+# FIXME add Altivec and AVX512.
+ISETS = (
+    'mmx',
+    'sse',
+    'sse2',
+    'sse3',
+    'ssse3',
+    'sse41',
+    'sse42',
+    'avx',
+    'avx2',
+    'neon',
+)
+
+
 class SimdModule(ExtensionModule):
 
     INFO = ModuleInfo('SIMD', '0.42.0', unstable=True)
 
-    def __init__(self, interpreter):
+    def __init__(self, interpreter: Interpreter):
         super().__init__(interpreter)
-        # FIXME add Altivec and AVX512.
-        self.isets = ('mmx',
-                      'sse',
-                      'sse2',
-                      'sse3',
-                      'ssse3',
-                      'sse41',
-                      'sse42',
-                      'avx',
-                      'avx2',
-                      'neon',
-                      )
         self.methods.update({
             'check': self.check,
         })
 
-    def check(self, state, args, kwargs):
-        result = []
-        if len(args) != 1:
-            raise mesonlib.MesonException('Check requires one argument, a name prefix for checks.')
-        prefix = args[0]
-        if not isinstance(prefix, str):
-            raise mesonlib.MesonException('Argument must be a string.')
-        if 'compiler' not in kwargs:
-            raise mesonlib.MesonException('Must specify compiler keyword')
+    @typed_pos_args('simd.check', str)
+    @typed_kwargs('simd.check',
+                  KwargInfo('compiler', Compiler, required=True),
+                  *[SOURCES_KW.evolve(name=iset) for iset in ISETS],
```

Comment:
Thinking about this more (because of the mypy issue), I think you should import the `STATIC_LIB_KWS` kwargs  as well and add something like:
```python
*[a for a in STATIC_LIB_KWS if a.name != 'sources'],
```
Which will mean that SIMD will automatically get any new values added to the static lib module for free, and that you could drop the `'if 'sources' in kwargs:` check below.

Reply:
Well, actually you can't drop the `if 'source'` until the `allow_unknown` is dropped...